capnp/lib/capnp.pyx

```diff
@@ -1771,51 +1760,150 @@ cdef class _DynamicObjectBuilder:
+    port.kjLoop.run()
+
+cdef void kjloop_advance_callback(void* data) with gil:
+    cdef AsyncIoEventPort *port = <AsyncIoEventPort*>data
+    assert port.runHandle is not None
+    port.timerImpl.advanceTo(systemPreciseMonotonicClock().now())
+
+cdef cppclass AsyncIoEventPort(EventPort):
+    EventLoop *kjLoop
+    TimerImpl *timerImpl;
+    object asyncioLoop;
+    object runHandle;
+
+    __init__(object asyncioLoop):
+        this.kjLoop = new EventLoop(deref(this))
+        this.timerImpl = new TimerImpl(systemPreciseMonotonicClock().now())
+        this.runHandle = None
+        this.asyncioLoop = asyncioLoop
+
+    __dealloc__():
+        del this.timerImpl
+        del this.kjLoop
+
+    cbool wait() except* with gil:
+        raise KjException("Currently you cannot wait for promises while pycapnp is running in asyncio mode. " +
+                          "You should instead use 'await'. If you have a use-case to start the asyncio loop " +
+                          "using wait(), please report")
+
+    cbool poll() except* with gil:
+        raise KjException("Currently you cannot poll promises while pycapnp is running in asyncio mode. " +
+                          "If you have a use-case to poll the asyncio loop using poll(), please report")
+
+    void setRunnable(cbool runnable) except* with gil:
+        if runnable:
+            if this.runHandle is not None:
+                # If a timer was running, cancel it and schedule a run immediately
+                # The timer will be re-scheduled once the kj loop becomes un-runnable again.
+                this.runHandle.cancel()
+            us = <void*>this;
+            this.runHandle = this.asyncioLoop.call_soon(lambda: kjloop_runnable_callback(us))
+        else:
+            assert this.runHandle is not None
+            this.runHandle.cancel()
+            this.scheduleAdvance()
+
+    void scheduleAdvance() with gil:
+        cdef uint64_t nextEvent = this.timerImpl.timeoutToNextEvent(
+            systemPreciseMonotonicClock().now(), MILLISECONDS, -1).orDefault(-1)
+        if nextEvent == -1:
+            this.runHandle = None
+        else:
+            seconds = <double>nextEvent / 1000
+            us = <void*>this;
+            this.runHandle = this.asyncioLoop.call_later(seconds, lambda: kjloop_advance_callback(us))
+
+    EventLoop *getKjLoop():
+        return this.kjLoop
+
+    Timer *getTimer():
+        return this.timerImpl;
+
+def _asyncio_close_patch(loop, oldclose, _EventLoop kjloop):
+    # The purpose of patching the asyncio close() function is to set up the kj-loop to be closed as well.
+    # We replace the event loop getter with a weakref, such that it can be destroyed when all other
+    # references to it are gone. Then, if a new asyncio loop ever gets started, a new kj-loop can also be
+    # started.
+    _C_DEFAULT_EVENT_LOOP_LOCAL.loop = _weakref.ref(kjloop)
+    loop.close = oldclose()
+    return oldclose()
```

Comment:
This seems to be causing an issue in my local tests. Should it not be `return oldClose`?

Reply:
Yes. Good eye.
I already have this fix as part of another PR. To be opened soon.